components/waypoint-movement.js

```diff
@@ -0,0 +1,239 @@
+      this.pathObject = pathObject;
+      this.children = this.pathObject.children.sort(function(a, b) { return a.name.localeCompare(b.name) });
+      this.positions = new Array(this.children.length);
+      for (var i = 0; i < this.children.length; i++) {
+        this.positions[i] = new Float32Array(3);
+        this.children[i].getTranslationWorld(this.positions[i]);
+      }
+      this.currentPositionIndex = 0;
+      this.currentCurveIndex = 0;
+      this.currentLength = 0;
+      this._setCurvePoints();
+
+      this.fromPosition = this.positions[0];
+      this.toPosition = this.positions[1];
+      glMatrix.vec3.sub(this.currentPosition, this.toPosition, this.fromPosition);
+      this.fromTolength = glMatrix.vec3.length(this.currentPosition);
+    },
+    update: function(dt) {
+      this.currentLength += dt * this.speed;
+      let factor = this.currentLength / this.fromTolength;
+      if(factor > 0.5 && this.currentPositionIndex != this.positions.length - 1) {
+        this.incrementCurveIndex = true;
+      }
+      if(Math.abs(factor - Math.round(factor)) < this.curveDistance && !(this.currentPositionIndex == this.children.length-2 && factor > 0.5) && !(this.currentPositionIndex == 0 && factor < 0.5)) {
+        let bezFactor = (factor - Math.round(factor) + this.curveDistance) * this.bezFactorMultiplicator;
+        glMatrix.vec3.bezier(this.currentPosition, this.p0, this.p1,this.p1, this.p2, bezFactor);
+        glMatrix.vec3.bezier(this.lookPosition, this.p0, this.p1,this.p1, this.p2, bezFactor+0.01);
+      } else {
+        if(this.incrementCurveIndex && factor < 0.5 && factor > this.curveDistance && this.currentCurveIndex != this.positions.length - 3) {
+          this.currentCurveIndex++;
+          this.incrementCurveIndex = false;
+          this._setCurvePoints();
+        }
+        glMatrix.vec3.lerp(this.currentPosition, this.fromPosition, this.toPosition, factor);
+        glMatrix.vec3.lerp(this.lookPosition, this.fromPosition, this.toPosition, factor+0.01);
+      }
+      this.object.resetTranslationRotation();
+      lookAt(this.quat, this.lookPosition, this.currentPosition, this.up);
+      this.object.rotate(this.quat);
+      this.object.setTranslationWorld(this.currentPosition);
+      if(this.currentLength > this.fromTolength){
+        this.start = false;
+        this.currentPositionIndex++
+        if(this.currentPositionIndex == this.children.length-1) {
+          this.currentPositionIndex = 0;
+          this.currentCurveIndex = 0;
+          this._setCurvePoints();
+          this.incrementCurveIndex = false;
+          this.onFinalWaypointReachedCallbacks.forEach(function(f){ f() });
+        }
+        this.fromPosition = this.positions[this.currentPositionIndex];
+        this.toPosition = this.positions[this.currentPositionIndex + 1];
+        glMatrix.vec3.sub(this.currentPosition, this.toPosition, this.fromPosition);
+        this.fromTolength = glMatrix.vec3.length(this.currentPosition);
```

Comment:
I think correct camel case would be `fromToLength`, but it's consistent everywhere, to not a problem